lib/services/unified_resource.go

```diff
@@ -173,9 +170,12 @@ func (c *UnifiedResourceCache) deleteLocked(res types.Resource) error {
 		return trace.NotFound("cannot delete resource: key %s not found in unified resource cache", key)
 	}
 
-	sortKey := makeResourceSortKey(resource)
-	c.deleteSortKey(sortKey)
-	delete(c.resources, key)
+	if resource.isSingular() {
+		c.deleteSortKey(resource.sortKey)
+		delete(c.resources, key)
+	} else {
+		resource.removeLocked(res)
+	}
```

Comment:
i don't know if it's possible but should we handle the case where `res` being removed is not part of cached list of `servers`?

Reply:
for server type, say database server, `isSingular` checks:
```
func (c *resourceCollection) isSingular() bool {
	return len(c.servers) <= 1
}
```
but is it possible if `c.servers[0]` is not the one getting removed here? say OpDelete happens twice for the 2nd last server. if it's unlikely then don't worry about it.